### ISSUE TITLE
Cnn multi-threading operation

### ISSUE THREAD

**Issue by LittleStoneLS** [tsogkas/brainseg#2]
Hi, 
There are some problems with my cuda environment, and it is hard to resolve it. So I want to try to run multithreaded on the CPU first. Could you please tell me if the project supports multithreading, and if possible what I need to do to implement it.

Thanks you very much!!

**Comment by LittleStoneLS** [comment]
Hi,

Please ignore the questions above. I have successfully run cnn.m on the GPU without cuDNN. But fail to run it with cuDNN~ The environment of my pc is ubuntu14.04+cuda8.0+cudnn6.0.
When I run

> vl_compilenn('enableGpu',true,...
'cudaRoot','/usr/local/cuda',...
'cudaMethod','nvcc',...
'enableCudnn',true,...
'cudnnRoot','/usr/local/cudnn-6.0')
it shows:

```
vl_compilenn:CUDA: MEX config file: '/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/config/mex_CUDA_glnxa64.xml'
Building with 'g++'.
MEX completed successfully.
Building with 'g++'.
MEX completed successfully.
Building with 'g++'.
MEX completed successfully.
/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/nnconv.cu(123): error: no instance of function template "vl::impl::nnconv_backward_cudnn" matches the argument list
            argument types are: (vl::Context, vl::Tensor, vl::Tensor, vl::Tensor, vl::Tensor, vl::Tensor, vl::Tensor, int, int, int, int, int, int)

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(37): error: no instance of function template "vl::impl::nnconv_forward_cudnn" matches the specified type

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(66): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(67): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(74): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(77): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(78): error: this declaration has no storage class or type specifier

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(79): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(90): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(91): error: this declaration has no storage class or type specifier

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(92): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(103): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(104): error: this declaration has no storage class or type specifier

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(105): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(112): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(125): warning: parsing restarts here after previous syntax error

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(126): error: this declaration has no storage class or type specifier

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(127): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(134): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(150): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(160): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(169): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(199): warning: parsing restarts here after previous syntax error

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(200): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(206): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(216): warning: parsing restarts here after previous syntax error

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(217): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(255): warning: parsing restarts here after previous syntax error

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(256): error: variable "error" has already been defined

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(257): error: variable "handle" has already been defined

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(260): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(263): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(264): error: this declaration has no storage class or type specifier

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(264): error: variable "dataDescInitialized" has already been defined

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(265): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(276): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(303): warning: parsing restarts here after previous syntax error

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(304): error: this declaration has no storage class or type specifier

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(304): error: variable "filtersDescInitialized" has already been defined

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(305): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(312): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(313): error: this declaration has no storage class or type specifier

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(313): error: variable "convDescInitialized" has already been defined

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(314): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(321): error: expected a declaration

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(370): warning: parsing restarts here after previous syntax error

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu(371): error: expected a declaration

40 errors detected in the compilation of "/tmp/tmpxft_00004f45_00000000-7_nnconv_cudnn.cpp1.ii".
Error using vl_compilenn>nvcc_compile (line 442)
Command "/usr/local/cuda/bin/nvcc" -c
"/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/nnconv_cudnn.cu" -DNDEBUG
-DENABLE_GPU -DENABLE_CUDNN -I/usr/local/cudnn-6.0 -I"/usr/local/MATLAB/R2014a/extern/include"
-I"/usr/local/MATLAB/R2014a/toolbox/distcomp/gpu/extern/include" -Xcompiler -fPIC
-gencode=arch=compute_61,code=\"sm_61,compute_61\"  -o
"/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/mex/.build/bits/impl/nnconv_cudnn.o" failed.

Error in vl_compilenn>(parfor body) (line 388)
      nvcc_compile(opts, srcs{i}, toobj(bld_dir,srcs{i}), flags.nvcc) ;

Error in vl_compilenn (line 384)
parfor i = 1:numel(horzcat(lib_src, mex_src))

```

There is also some warning information when running

> vl_compilenn('enableGpu',true,...
'cudaRoot','/usr/local/cuda',...
'cudaMethod','nvcc')

it shows:

```
In file included from /usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/imread_libjpeg.cpp:14:0:
/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/imread_helpers.hpp:35:2: warning: #warning "SSSE3 instruction set not enabled. Using slower image conversion routines." [-Wcpp]

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/bits/impl/tinythread.h(654): warning: extra ";" ignored

/usr/local/MATLAB/Matlab_project_file/brainseg-master/MatConvNet/matlab/src/vl_imreadjpeg.cu(223): warning: variable "num" was declared but never referenced
```

I don‘t know what the problem is. Help if possible.
Thank you very much!!

LS

**Comment by tsogkas** [comment]
I am sorry but this code has not been tested using cudnn. I have modified the convolution routines in MatConvNet to perform "dilated" convolution, but that does not involve support for cudnn (when I was working on that project dilated convolutions were not yet implemented in MatConvNet).

You could perhaps try to use a more recent version of MatConvNet if you really want to use cudnn.

Hope that helps.

**Comment by LittleStoneLS** [comment]
OK, I‘ll try.  
Thank you.

In addition, I wonder how long your program is running at that time. Mine has been run for more than three days~

**Comment by tsogkas** [comment]
Yes in my case it was taking 2-3 days as well.

EDIT: it depends on how long you wish to train, 2-3 days corresponded to ~25-30 full passes of the training dataset if I remember correctly.

**Comment by mosbahi1** [comment]
Hello
when I run this code the output was a file.mat in  the file result/cnn but
I don't find the image segmented

2017-08-09 4:48 GMT+01:00 Stavros Tsogkas <notifications@github.com>:

**Comment by tsogkas** [comment]
@mosbahi1 This code is for training a CNN model, which is saved in a .mat file. If you want to test on an image you have to run a trained model on the input image and then extract segmentation labels from the CNN probabilistic outputs (using a softmax layer).

PS: you should start a new thread for issues that are not related to the current thread.

**Comment by mosbahi1** [comment]
thank you
I add a softamax layer, then I run this training, then I extract
segmentation labels from the output of sftamax layer

2017-08-10 2:14 GMT+01:00 Stavros Tsogkas <notifications@github.com>:

**Comment by mosbahi1** [comment]
I dont know how I add softmax in the code


2017-08-10 8:08 GMT+01:00 mosbahi sourour <mosbahisourour@gmail.com>:

**Comment by LittleStoneLS** [comment]
Thanks.
